api/v1alpha1/common_test.go

```diff
@@ -0,0 +1,314 @@
+/*
+Copyright 2023 The Kubebb Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"reflect"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type condStatusTestCase struct {
+	status, exp ConditionedStatus
+	l           int
+	cond        Condition
+}
+
+func TestUpdateCondwithFixedLen(t *testing.T) {
```

Comment:
just for fun. This is the unit test generate by codeium
```golang
func TestUpdateCondWithFixedLen(t *testing.T) {
    testCases := []struct {
        name            string
        l               int
        input           *ConditionedStatus
        cond            Condition
        expectedOutput  *ConditionedStatus
    }{
        {
            name: "Update with l = 0",
            l: 0,
            input: &ConditionedStatus{
                Conditions: []Condition{
                    {
                        Type: "Ready",
                        Status: "True",
                    },
                },
            },
            cond: Condition{
                Type: "Progressing",
                Status: "False",
            },
            expectedOutput: &ConditionedStatus{
                Conditions: []Condition{
                    {
                        Type: "Ready",
                        Status: "True",
                    },
                    {
                        Type: "Progressing",
                        Status: "False",
                    },
                },
            },
        },
        {
            name: "Update with l < len(input.Conditions)",
            l: 1,
            input: &ConditionedStatus{
                Conditions: []Condition{
                    {
                        Type: "Ready",
                        Status: "True",
                    },
                    {
                        Type: "Progressing",
                        Status: "True",
                    },
                },
            },
            cond: Condition{
                Type: "Progressing",
                Status: "False",
            },
            expectedOutput: &ConditionedStatus{
                Conditions: []Condition{
                    {
                        Type: "Progressing",
                        Status: "True",
                    },
                    {
                        Type: "Progressing",
                        Status: "False",
                    },
                },
            },
        },
        {
            name: "Update with l > len(input.Conditions)",
            l: 3,
            input: &ConditionedStatus{
                Conditions: []Condition{
                    {
                        Type: "Ready",
                        Status: "True",
                    },
                },
            },
            cond: Condition{
                Type: "Progressing",
                Status: "False",
            },
            expectedOutput: &ConditionedStatus{
                Conditions: []Condition{
                    {
                        Type: "Ready",
                        Status: "True",
                    },
                    {
                        Type: "Progressing",
                        Status: "False",
                    },
                },
            },
        },
    }

    for _, tc := range testCases {
        t.Run(tc.name, func(t *testing.T) {
            UpdateCondWithFixedLen(tc.l, tc.input, tc.cond)
            if !reflect.DeepEqual(tc.input, tc.expectedOutput) {
                t.Errorf("Expected output to be %v, but got %v", tc.expectedOutput, tc.input)
            }
        })
    }
}

```

this is unit test generate by chatgpt
```golang
import (
	"reflect"
	"testing"
)

func TestUpdateCondWithFixedLen(t *testing.T) {
	// Define test cases as a table
	tests := []struct {
		name      string
		inputLen  int
		inputConds *ConditionedStatus
		inputCond Condition
		expected  *ConditionedStatus
	}{
		{
			name:      "Empty list with non-zero length",
			inputLen:  5,
			inputConds: &ConditionedStatus{Conditions: []Condition{}},
			inputCond: Condition{Type: "A", Value: 1},
			expected: &ConditionedStatus{Conditions: []Condition{{Type: "A", Value: 1}}},
		},
		{
			name:      "List with length less than fixed length",
			inputLen:  5,
			inputConds: &ConditionedStatus{Conditions: []Condition{{Type: "A", Value: 1}, {Type: "B", Value: 2}}},
			inputCond: Condition{Type: "C", Value: 3},
			expected: &ConditionedStatus{Conditions: []Condition{{Type: "A", Value: 1}, {Type: "B", Value: 2}, {Type: "C", Value: 3}}},
		},
		{
			name:      "List with length equal to fixed length",
			inputLen:  2,
			inputConds: &ConditionedStatus{Conditions: []Condition{{Type: "A", Value: 1}, {Type: "B", Value: 2}}},
			inputCond: Condition{Type: "C", Value: 3},
			expected: &ConditionedStatus{Conditions: []Condition{{Type: "B", Value: 2}, {Type: "C", Value: 3}}},
		},
		{
			name:      "List with length greater than fixed length",
			inputLen:  2,
			inputConds: &ConditionedStatus{Conditions: []Condition{{Type: "A", Value: 1}, {Type: "B", Value: 2}, {Type: "C", Value: 3}}},
			inputCond: Condition{Type: "D", Value: 4},
			expected: &ConditionedStatus{Conditions: []Condition{{Type: "C", Value: 3}, {Type: "D", Value: 4}}},
		},
	}

	// Iterate over the test cases
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			// Call the function under test
			UpdateCondWithFixedLen(test.inputLen, test.inputConds, test.inputCond)

			// Check if the conditions list matches the expected result
			if !reflect.DeepEqual(test.inputConds, test.expected) {
				t.Errorf("Expected: %v, Got: %v", test.expected, test.inputConds)
			}
		})
	}
}

```